RAR as service design (#5554)
dotnet/msbuild: documentation/specs/rar-as-service.md

# Resolve Assembly Reference as Service Design

This document describes the Resolve Assembly Reference as Service

# Background

> [MSBuild](https://docs.microsoft.com/en-us/visualstudio/msbuild/msbuild?view=vs-2019) is a universal build engine, used for building pretty much everything in the Microsoft world. It is available on command line (msbuild, [dotnet build](https://docs.microsoft.com/en-us/dotnet/core/tools/dotnet-build)), runs under the covers when building projects and solutions in Visual Studio, and is used as the local build engine in "higher-order" distributed build systems. Essentially all .NET applications use MSBuild as their primary build engine.

Author: Most of the background section was taken from @ladipro document, where he introduced me this project for my internship. I have had link to that (and still is in the docx version, albeit that version is now little outdated), but removed it since that document is internal. 

I don't want to suggest what he thought by that, but how I explain it is, that MSBuild can be used for example with [QuickBuild](https://www.pmease.com/).

> [MSBuild](https://docs.microsoft.com/en-us/visualstudio/msbuild/msbuild?view=vs-2019) is a universal build engine, used for building pretty much everything in the Microsoft world. It is available on command line (msbuild, [dotnet build](https://docs.microsoft.com/en-us/dotnet/core/tools/dotnet-build)), runs under the covers when building projects and solutions in Visual Studio, and is used as the local build engine in "higher-order" distributed build systems. Essentially all .NET applications use MSBuild as their primary build engine.

Author: I have changed the docs links, even though when I get to the docs and they are in my local language I always switch them to English. :)

> RAR is the acronym behind ResolveAssemblyReference (an MSBuild task) and ResolveAssemblyReferences (an MSBuild target). RAR is used in all .NET builds. Quoting the official documentation, RAR "_Determines all assemblies that depend on the specified assemblies, including second and nth-order dependencies._"

Author: Will update. 

According to this document yes. https://github.com/dotnet/project-system/blob/master/docs/design-time-builds.md#targets-that-run-during-design-time-builds. It is 1.5 year old, but I don't think that there were any changes to this logic.

The RAR task has become very complex and slow over the years. It tends to rank high on the list of MSBuild's performance bottlenecks. There is an inherent cost to walking the assembly reference graph and computing the dependency closure, especially in terms of I/O operations. To address this, the task internally maintains caches, both in-memory and on disk. While it alleviates the problem somewhat, it is still a suboptimal solution because

1. The task runs in build nodes which are generally created as separate processes, one per logical CPU, so the in-memory state is duplicated.
2. Build nodes have limited lifetime and the in-memory state is lost when they die.
3. No state exists when the task runs for the given project for the first time.

Point 3 is tracked by issue [#5247](https://github.com/dotnet/msbuild/issues/5247).

> There was already an attempt to introduce this to MSBuild ([#3914](https://github.com/dotnet/msbuild/pull/3914)). This PR was not completed mainly because of discontinued development of Bond, which is in that PR used as method of communication between nodes.

Author: Fixed

# Design

![](assets/rar-lifetime.png)

_Figure 1 Rough diagram of lifetime of service_

## Lifetime of service

### Connect to RAR node

Connecting to the RAR node will not require any discovery of processes on computer. The algorithm will follow these steps:

1. Get expected node name, which will be based on the current setting of node. The name format is described in `Start new node`.
2. Setup named pipe. The construction of named pipe may differ between platforms (usage of different API for construction of the pipe object).

> 3. Try to connect to the node.

Author: I don't want to modify the handshake. I will just reuse the part of code which figures out the HnadshakeOption and form that I will construct the "setting number" in name of node. In my original design was that we would also do the normal handshake procedure, but Rainer pointed out that the handshake here is already encoded in the name of node se we can rely on it.

If the connection is successful, we can use this connection for execution of RAR task. The node is guaranteed to have all required properties since they must be encoded in name of the RAR node.

### Start new node

> This step will create new process which will act as RAR node. It will also pass necessary information to the node to know what its settings are (reusable, ...). Node will be another instance of the MSBuild.exe which will have set parameter **nodeMode** to some specific value. Probably it will be text value, so we can easily recognize which node is RAR node and which is normal MSBuild node.

Author: It should be nodeMode 3. There are cases for nodeMode 1 and 2. I guess you can take normal mode as nodeMode 0 (but definitely not use it). So 4 in total.

The RAR node, also has to support accepting of already established commands for MSBuild nodes (for example Shutdown command). This will be done by creating two pipes inside node, one will be for communication about RAR commands and second one for the servicing communication. 

### Execute task in MSBuild node

User opted out of using the RAR nodes so we will execute the RAR task in the MSBuild node (as it is right now).

### Other

The new RAR node will not count toward total maximum CPU count specified by _/maxCpuCount_ switch, since the RAR task is taxing on IO operations not so much on CPU time. If we took one node from each instance of MSBuild it would lead to drastic decrease in performance.

The RAR task will be affected by the _/m_ switch. When we run in single node mode, it will implicitly say that we want to run the task inside current process. User would have to explicitly say that they want to use the RAR node.

__NOTE:__ The behavior described above depend on fact that the feature is opt-out (is active by default). If not, the paragraph above is meaningless. This has to be yet decided/clarified. 

## Communication

> The communication between nodes should be done over [StreamJsonRpc](https://github.com/microsoft/vs-streamjsonrpc/). The API over which two node will transfer data has to reflect inputs and outputs of RAR task as described in [docs](https://docs.microsoft.com/en-us/visualstudio/msbuild/resolveassemblyreference-task?view=vs-2019). It is possible that into the API may be added some additional data.

Reviewer: Again, fix link

I don't think the last sentence is necessary, but if you do:
`We may augment the API with more data.` sounds better.

RAR is the acronym behind ResolveAssemblyReference (an MSBuild task) and ResolveAssemblyReferences (an MSBuild target). RAR is used in all .NET builds. Quoting the official documentation, RAR "_Determines all assemblies that depend on the specified assemblies, including second and nth-order dependencies._"

The RAR task has become very complex and slow over the years. It tends to rank high on the list of MSBuild's performance bottlenecks. There is an inherent cost to walking the assembly reference graph and computing the dependency closure, especially in terms of I/O operations. To address this, the task internally maintains caches, both in-memory and on disk. While it alleviates the problem somewhat, it is still a suboptimal solution because

1. The task runs in build nodes which are generally created as separate processes, one per logical CPU, so the in-memory state is duplicated.

> 1. Build nodes have limited lifetime and the in-memory state is lost when they die.

Reviewer 2: The numbering is wrong here.

Author: Thanks

3. No state exists when the task runs for the given project for the first time.

> Nathan Mytelka is working on 3. He is planning to ship pre-computed caches as part of VS/SDK so even clean build will be able to skip walking the reference graph for at least system assemblies.

Reviewer 3: I'd remove Nathan's name and either just say that it's being worked on separately, or link to an issue/PR.

### Execute RAR task

Execution of RAR task should be done in same way as it is done now.

There is already some layer of separation between Task interface and actual execution method. We will leverage this, and put the decision logic if to run locally or not into the "wrapper" method and so we will not have to modify this and in server-side execution we will directly call the internal execution method.

There is one big concern and that is how to handle multiple requests at once. As right now, RAR task is not prepared for multi-thread use.

### Shutdown RAR node

> If the user want's the node to not be reused, we have the ensure that node will be killed after the build ends. This should be done after the compilation end by main MSBuild node.

Reviewer 3: nit: `wants`

> The RAR node, also has to support accepting of already established commands for MSBuild nodes (for example Shutdown command). This will be done by creating two pipes inside node, one will be for communication 

Reviewer 3: Looks like an unfinished sentence.

### Execute task in MSBuild node

User opted out of using the RAR nodes so we will execute the RAR task in the MSBuild node (as it is right now).

### Other

The new RAR node will not count toward total maximum CPU count specified by _/maxCpuCount_ switch, since the RAR task is taxing on IO operations not so much on CPU time. If we took one node from each instance of MSBuild it would lead to drastic decrease in performance.

> The RAR task will be affected by  _/m_ switch. When we run in single node mode, it will implicitly say that we want to run task inside current process. User would have to explicitly say that he wants to run with RAR node.

Reviewer 3: ```suggestion
The RAR task will be affected by the _/m_ switch. When we run in single node mode, it will implicitly say that we want to run the task inside current process. User would have to explicitly say that they want to use the RAR node.
```

And a question: Assuming that the feature is going to be opt-in, even after it's fully implemented, isn't the above statement true regardless of _/m_? I.e. if not opted in, it's going to behave just like today, running the task in the process building the respective project, and if opted in it will use the RAR node always?

Author: Text updated. That paragraph is valid if the feature is opt-out. Added note, that the default state of this feature is not yet decided.

    public Microsoft.Build.Framework.ITaskItem[] SuggestedRedirects { get { throw null; }
}
```
### Logging response

Node also has to return logged data, preferably in stream which will be transferred during the RAR task execution to the client node. This should by done by using IProgress\<T> support in StreamJsonRPC. There is also support for async enumerable, but they must be pulled and are not pushed to the client (comparison can be found [here](https://github.com/microsoft/vs-streamjsonrpc/blob/master/doc/asyncenumerable.md#comparison-with-iprogresst)).

### Concurrency

One of the biggest challenges with RAR as service, is to make execution and caching of RAR task thread-safe, since in most cases there will be multiple clients request data from it at once.

In the first phase of implementation the concurrency will be solved by serializing requests, but this will be replaced by optimal variant with no serializing.

# Non-Goals

- File watchers: using them would decrease required IO operations when checking disc changes
- Aggressive precomputation of results
- Improved caching of requests

> - Providing verbosity to RAR task

Reviewer 3: Thank you for explicitly listing it as a non-goal. Should the text about logging under "RAR Input", as well as the `Verbosity` property in `ResolveAssemblyReferenceInput` be deleted, then?
